### ISSUE TITLE
Plugin does not work with application load balancers

### ISSUE THREAD

**Issue by rhunwicks** [shinsaka/fluent-plugin-elb-log#22]
Amazon now has Application Load Balancers as well as Classic Load Balancers. Application Load Balancers have a slightly different set of fields in the logs saved to S3 - see https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html?icmpid=docs_elbv2_console vs http://docs.aws.amazon.com/elasticloadbalancing/latest/classic/access-log-collection.html.

Specifically application load balancers have `type` at the start of the line and `target_group_arn` and `trace_id` at the end.

The following regex works for both classic and application load balancers as far as I can tell:

```ruby
ACCESSLOG_REGEXP = /^((?<type>[a-z0-9]+) )?(?<time>\d{4}-\d{2}-\d{2}T\d{2}\:\d{2}\:\d{2}\.\d{6}Z) (?<elb>.+?) (?<client>[^ ]+)\:(?<client_port>.+?) (?<backend>.+?)(\:(?<backend_port>.+?))? (?<request_processing_time>.+?) (?<backend_processing_time>.+?) (?<response_processing_time>.+?) (?<elb_status_code>.+?) (?<backend_status_code>.+?) (?<received_bytes>.+?) (?<sent_bytes>.+?) \"(?<request_method>.+?) (?<request_uri>.+?) (?<request_protocol>.+?)\"( \"(?<user_agent>.*?)\" (?<ssl_cipher>.+?) (?<ssl_protocol>[^ ]+)( (?<target_group_arn>arn:.+) (?<trace_id>.+))?(| (?<option3>.*)))?/
```

Also, the files are gzipped and have a `.log.gz` extension. The follow regex matches both the Classic and the Application Load Balancer file names:

```
LOGFILE_REGEXP = \^((?<prefix>.+?)\/|)AWSLogs\/(?<account_id>[0-9]{12})\/elasticloadbalancing\/(?<region>.+?)\/(?<logfile_date>[0-9]{4}\/[0-9]{2}\/[0-9]{2})\/[0-9]{12}_elasticloadbalancing_.+?_(?<logfile_elb_name>[^_]+)_(?<elb_timestamp>[0-9]{8}T[0-9]{4}Z)_(?<elb_ip_address>.+?)_(?<logfile_hash>.+)\.log(.gz)?$\
```

Unfortunately, we would still need to uncompress the files after we have retrieved them. I'm not a Ruby programmer, but I am assuming something like:

```ruby
  def get_file_from_s3(object_name)
    begin
      log.debug "getting object from s3 name is #{object_name}"

      # read an object from S3 to a file and write buffer file
      File.open(@buf_file, File::WRONLY|File::CREAT|File::TRUNC) do |file|
        s3_client.get_object(
          bucket: @s3_bucketname,
          key: object_name
        ) do |chunk|
          if object_name[-3, 3] == '.gz'
            sio = StringIO.new(chunk)
            gz = Zlib::GzipReader.new(sio)
            data = gz.read
            gz.close
            file.write(data)
          else
            file.write(chunk)
          end
        end
      end
    rescue => e
      log.warn "error occurred: #{e.message}"
    end
  end
```

**Comment by shinsaka** [comment]
@rhunwicks Thank you for your comment.I have begun this issue.
I'm sorry for my lated reply.

**Comment by shinsaka** [comment]
I released gem version 0.4.0.
Could you try new version.

Thank you for your advice.
